### ISSUE TITLE
use Moodle 2.7's new Event API

### ISSUE THREAD

**Issue by ndunand** [ndunand/moodle-tool_mergeusers#50]
events_trigger() is still working in Moodle 2.7 but has been deprecated
see http://docs.moodle.org/dev/Event_2#Events_API

**Comment by MorrisR2** [comment]
I started work on this, but then I remembered @jpahullo recently made a commit to keep mergeusers compatible with Moodle 2.4.  New events require Moodle 2.6 or higher, I think.  Do we want to do this now, or wait?

**Comment by jpahullo** [comment]
Hi all!

I think we need to start working on several branches (not only master, but MOODLE_24_STABLE and so on) as suggested in the comments from September Featured plugin.

We talked about it privately with @ndunand. Do you remember? What do you think @ndunand ?

**Comment by ndunand** [comment]
Well, two thoughts:
 1. do we still want to support legacy (i.e., unsupported) Moodle versions?
 2. do we want to go through the extra overhead of maintaining a separate branch for each of these?

My opinion on 1. is that we should restrict support to current Moodle versions. About 2., I'm not sure, but wouldn't it be simpler to check against the Moodle version in the code?

What do you think?

**Comment by MorrisR2** [comment]
It seems to me that checking against the Moodle version within the code would be easier *in the short term*.  Five years from now, we'd have a bunch of version conditions littered throughout the code, for very old versions that nobody is using. That could get quite messy over time. Separate branches allow us to drop or freeze Moodle 2.4 (or any other version) support when we choose to do so, simply by not updating the branch any more.  That seems better than needing to dig through the code and excise the stuff that is specific to each deprecated version. 

Git makes it easy to apply a commit in one branch to the other branches.  If we didn't use git, multiple branches would be more difficult. 

As to #1, which versions to support, I guess @jpahullo had a *reason* to support 2.4 - he's either using 2.4 or knows somebody that uses the tool with 2.4.  In the kernel, support for older stuff is maintained if and only if people are using it. If we know of at least one institution using this tool with 2.4, we should keep that in mind. Perhaps backport at least fixes for significant bugs.  If we don't know of anyone using 2.4, that's also instructive.

**Comment by ndunand** [comment]
Well, we'd hardly be checking against more than a few (2?) concurrent (supported) Moodle versions, but I see your point. However, how would we handle branches? (assuming we formally decide to support unsupported Moodle versions). Let's say for example there's no change between M2.4-M2.6, so would we still keep three branches (_24, _25, _26) for clarity? If so, and if we still have to support Moodle 2.4, then soon we'd have to support 5 different branches.

For me the main question remains: which Moodle versions to support? @jpahullo , is Moodle 2.4 support still necessary, or can we drop it and focus on "current" Moodle versions?

In any way, the overhead of supporting numerous Moodle versions will be the same either way (branches or code checks). If we only ever support 2 Moodle versions, then code checking would remain the best solution.

Open to discussion...

**Comment by MorrisR2** [comment]
I'd think we'd branch when needed - if we use a feature that is available from 3.2 onward, we branch at that point.  So the branch would be based on "requires Moodle x.y or higher".

> In any way, the overhead of supporting numerous Moodle versions will be the same either way

I guess I was unclear.  Removing support for a version, with checks in the code:
Find the parts that check for versions, like this:
```
if ($moodleversion > 3.0) {
    dosomething(data);
```

THEN look through and see if anything else calls dosomething(), or if dosomething() can be removed.

Transitioning to  new version by using branches:
git checkout -b mdl_32

Dropping support for the old version consists of - **doing nothing**.  Just stop updating the mdl_28 branch.

**Comment by jpahullo** [comment]
Hi all!

I was smiling a bit while reading your comments. Don't bother me. It was because all this stuff was discussed by Nicolas and me in private messages, and the conclusions from both of you are becoming again the same :smile: 

In my honest opinion, considering all what we have discussed before, we should proceed like this:

1. Branching instead of version controlling. This is the way Moodle proceeds. I don't like some things done in Moodle, but this, powered by git, becomes very easy to maintain over time.
1. We will use Moodle 2.4 still until June/July 2016. Some hardware restrictions in our institution forced this decision.

What about branching? A branch per Moodle version or a branch per set of Moodle versions? Nicolas and me widely talked about this, and for our simplicity, a branch per set of Moodle versions was prominent. However, it breaks the way Moodle works and make confuse people that will be tracking our github repository. 

In conclusion, I would like to proceed like in Moodle with a little bit of improvement:

1. A branch per Moodle version. By now, all MOODLE_xx_STABLE (starting from 2.4) would be the same like our master.
2. Bug solutions, when applicable, should be included for legacy versions too.

The effort for behaving like this would be for the person making the PR (which will have to make one per applicable branch). So, you @ndunand, would have little work :stuck_out_tongue: 

In my case, for me in particular, if I will be doing some improvement, I'll design it starting from 2.4.

What do you think?

Jordi

**Comment by danmarsden** [comment]
in case it's useful to hear what I do....

I only maintain branches that require a difference. - the master branch is used for all the latest versions until I need branch for a specific version. This means that I might have:
MOODLE_23_STABLE
MOODLE_26_STABLE
master

I've occasionally had people ask me to change that so they can use submodules to track the same branch as their main branch but my response is usually "I maintain this plugin as a volunteer, maintaining a smaller number of branches is simpler/faster for me which gives me more time for more important stuff." 

I guess if more people start making wider use of submodules to track 3rd party plugins I might re-consider that at some point.

Branching is definitely the way to go rather than adding hacks/workarounds to detect various moodle versions and act as required. I've seen a lot of 3rd party plugins that try to support multiple Moodle versions in a single plugin codebase and it just gets really nasty to support/maintain in the long term.

**Comment by jpahullo** [comment]
Thanks @danmarsden, @ndunand, @MorrisR2.

It seems that definitive solution is having branches, with a branch per set of Moodle versions (like "MOODLE_23_STABLE, MOODLE_26_STABLE, master"). I like it too.

The final decision, though, is yours @ndunand :+1: 

Jordi

**Comment by ndunand** [comment]
Thanks all for sharing your thoughts. Indeed it seems reasonable to have branching on an "as needed" basis. Should we then have
 - MOODLE_24_STABLE for @jpahullo needing something to work on 2.4
 - do we need a specific branch for Moodle 2.5? (i.e. without the new Events system)
 - master

**Comment by jpahullo** [comment]
Rereading all what we've talked about, I think we would need MOODLE_24_STABLE (for all we who use M2.4, many thanks and sorry), MOODLE_27_STABLE (for adapting this plugin to the new events API, as stated in this subject's issue) and master. master should point to the last MOODLE_XX_STABLE branch, shouldn't it?

Thanks!

**Comment by ndunand** [comment]
All right, let's go for that.
